<a href="https://colab.research.google.com/github/Justus-coded/Competitions/blob/master/DSN%20AI%20Bootcamp%20Qualification%20Hackathon/DSN%20AI%20Boot_camp%20Qualification1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
 
%matplotlib inline
 
import warnings
warnings.filterwarnings('ignore')

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
train = pd.read_csv('/content/drive/My Drive/DSN AI Bootcamp Qualification Hackathon/Train.csv')
test  = pd.read_csv('/content/drive/My Drive/DSN AI Bootcamp Qualification Hackathon/Test.csv')
 
#train = pd.read_csv('../input/dsn-ai-bootcamp-data/Train.csv')
#test = pd.read_csv('../input/dsn-ai-bootcamp-data/Test.csv')
#sub = pd.read_csv('../input/dsn-ai-bootcamp-data/SampleSubmission.csv')

In [4]:
dict_ = {'charge': 1,
         'lending': 0}
 
target_map = {
    'yes': 1,
    'no': 0}
 
train.default_status.replace(target_map, inplace = True)
 
train['form_field47'].replace(dict_, inplace = True)
test['form_field47'].replace(dict_, inplace = True)

In [5]:
pip install catboost

     |████████████████████████████████| 66.1MB 87kB/s 


In [6]:
from sklearn.metrics import roc_auc_score
from catboost import CatBoostClassifier

In [7]:
features = train.select_dtypes(exclude = object).columns.drop(['default_status'])

In [8]:
# features

In [9]:
for col in features:
    train[col].fillna(-999, inplace = True)
    test[col].fillna(-999, inplace = True)

In [32]:
train['form_field1*formfield2'] = train['form_field1'] * train['form_field2']
train['form_field1/formfield2'] = train['form_field1'] / train['form_field2']
 
train['form_field1*formfield3'] = train['form_field1'] * train.form_field3
train['form_field1*formfield4'] = train.form_field1 * train.form_field4
train['form_field1*formfield5'] = train.form_field1 * train.form_field5
 
#train['form_field30*formfield41'] = train['form_field30'] * train['form_field41']
#train['form_field31*formfield40'] = train['form_field31'] * train['form_field40']
 
train['form_field6+field7'] =  (train.form_field6 + train.form_field8)/2
 
train['form_field1*formfield6'] = train['form_field1'] * train['form_field6']
train['form_field2*formfield6'] = train['form_field2'] * train['form_field6']
 
train['form_field1*formfield9'] = train['form_field1'] * train['form_field9']
train['form_field2*formfield9'] = train['form_field2'] * train['form_field9']
 
train['form_field1*formfield37'] = train['form_field1'] * train['form_field37']
train['form_field2*formfield37'] = train['form_field2'] * train['form_field37']
 
train['form_field9*formfield6'] = train['form_field9'] * train['form_field6']
train['form_field37*formfield6'] = train['form_field37'] * train['form_field6']
 
train['form_field9*formfield37'] = train['form_field9'] * train['form_field37']
train['form_field1*formfield8'] = train['form_field1'] * train['form_field8']
 
 
 
 
#train['form_field13/field14'] =(train.form_field13 / train.form_field14)*100
 
#train['form_field32/field33'] = (train['form_field32']+ train['form_field33'])/2
#train['form_field36/field38'] = train['form_field36'] / train['form_field38']
#train['form_field36/field39'] = train['form_field36'] / train['form_field39']
#train['form_field36/field18'] = train['form_field36'] / train['form_field18']
#train['form_field36/field20'] = train['form_field36'] / train['form_field20']
#train['form_field44*field50'] = train['form_field44'] * train['form_field50']

In [33]:
test['form_field1*formfield2'] = test['form_field1'] * test['form_field2']
test['form_field1/formfield2'] = test['form_field1'] / test['form_field2']
 
test['form_field1*formfield3'] = test['form_field1'] * test.form_field3
test['form_field1*formfield4'] = test.form_field1 * test.form_field4
test['form_field1*formfield5'] = test.form_field1 * test.form_field5
 
#test['form_field30*formfield41'] = test['form_field30'] * test['form_field41']
#test['form_field31*formfield40'] = test['form_field31'] * test['form_field40']
 
test['form_field6+field7'] =  (test.form_field6 + test.form_field8)/2
 
test['form_field1*formfield6'] = test['form_field1'] * test['form_field6']
test['form_field2*formfield6'] = test['form_field2'] * test['form_field6']
 
test['form_field1*formfield9'] = test['form_field1'] * test['form_field9']
test['form_field2*formfield9'] = test['form_field2'] * test['form_field9']
 
test['form_field1*formfield37'] = test['form_field1'] * test['form_field37']
test['form_field2*formfield37'] = test['form_field2'] * test['form_field37']
 
test['form_field9*formfield6'] = test['form_field9'] * test['form_field6']
test['form_field37*formfield6'] = test['form_field37'] *test['form_field6']
 
test['form_field9*formfield37'] = test['form_field9'] * test['form_field37']
test['form_field1*formfield8'] = test['form_field1'] * test['form_field8']
 
 
 
#test['form_field13/field14'] = (test.form_field13 / test.form_field14)*100
#test['form_field32/field33'] = (test['form_field32']+ test['form_field33'])/2
#test['form_field36/field38'] = test['form_field36'] / test['form_field38']
#test['form_field36/field39'] = test['form_field36'] / test['form_field39']
#test['form_field36/field18'] = test['form_field36'] / test['form_field18']
 
#test['form_field36/field20'] = test['form_field36'] / test['form_field20']
#test['form_field44*field50'] =test['form_field44'] * test['form_field50']

In [34]:
 for col in features: 
   
 
    train[col + '_mean_on_field1_'] = train[col].groupby([train['form_field47'],train['form_field1']]).transform('mean')
    train[col + '_mean_on_field2_'] = train[col].groupby([train['form_field47'],train['form_field2']]).transform('mean')
 
    train[col + '_sum_on_field1_'] = train[col].groupby([train['form_field47'],train['form_field1']]).transform('sum')
    train[col + '_sum_on_field2_'] = train[col].groupby([train['form_field47'],train['form_field2']]).transform('sum')

    test[col + '_mean_on_field1_'] = test[col].groupby([test['form_field47'],test['form_field1']]).transform('mean')
    test[col + '_mean_on_field2_'] = test[col].groupby([test['form_field47'],test['form_field2']]).transform('mean')

    test[col + '_sum_on_field1_'] = test[col].groupby([test['form_field47'],test['form_field1']]).transform('sum')
    test[col + '_sum_on_field2_'] = test[col].groupby([test['form_field47'],test['form_field2']]).transform('sum')

In [50]:
num_cols = train.select_dtypes(exclude = object).columns.drop(['default_status'])
len(num_cols), len(features)


(266, 50)

In [41]:
#features

In [37]:
X = train[num_cols]
y = train['default_status']

In [47]:
test_cols = test.select_dtypes(exclude = object).columns


In [14]:
from sklearn.model_selection import StratifiedKFold

In [15]:
def metric(y, pred):
    return roc_auc_score(y, pred, labels=[0, 1])

In [42]:
# Specify number of folds
n_skf = 15
kf = StratifiedKFold(n_skf)
 
#seed = 42
 
## Manually tweak the hyperparameter, learning rate and n_estimators especially
params = {'max_depth':7,
    'n_estimators': 5000,
    'learning_rate': 0.008,
    'objective': 'CrossEntropy',
      'eval_metric':'AUC',
    
    'early_stopping_rounds': 200,
    'use_best_model': True,
}

In [ ]:
score_list = []
score = 0
test_oofs = []
 
for i, (tr_idx, vr_idx) in enumerate(kf.split(X, y)):
    
    xtrain, ytrain = X.loc[tr_idx, num_cols], y.loc[tr_idx]
    xval, yval = X.loc[vr_idx, num_cols], y.loc[vr_idx]
 
    model = CatBoostClassifier(**params, task_type='GPU',random_seed = i)
    model.fit(xtrain, ytrain, eval_set=[(xval,yval)], verbose=0)
    
    p = model.predict_proba(xval)[:, 1]
    sc = metric(yval, p)
    score_list.append(sc)
    score += sc/n_skf
    
    pred = model.predict_proba(test[num_cols])[:, 1]
    test_oofs.append(pred)
 
    print('Fold {} : {}'.format(i, sc))
 
print()
print()
print('Avg log : ', score)

Fold 0 : 0.8284280942937948


In [18]:
f"{n_skf} fold CV, score: {score}" #Increase in local cv score should probably lead to increase in Leaderboard score

'15 fold CV, score: 0.8419128124711005'

15 fold CV, score: 0.8419128124711005, LR-0.008, n_estimator = 5000, seed = i, depth = 7, submit12.csv, Added more aggregate features

'15 fold CV, score: 0.8416762360251805', LR-0.005, n_estimator = 10000, seed = i, depth = 7, submit10.csv

'15 fold CV, score: 0.8416215384358319', LR-0.008, n_estimator = 5000, seed = i, depth = 8

'20 fold CV, score: 0.8419991965085709',LR-0.008, n_estimator = 5000, seed = i, depth = 7, submit9.csv

'15 fold CV, score: 0.8419018702724905', LR-0.008, n_estimator = 5000, seed = i, depth = 7, submit8.csv, Best score on LB

'15 fold CV, score: 0.8418325697189986, LR-0.008, n_estimator = 5000, seed = 42, depth = 7, submit7.csv

'10 fold CV, score: 0.8416283637107365', LR-0.008, n_estimator = 5000, seed = 42, depth = 7, submit5.csv

'10 fold CV, score: 0.8404761359972677', LR-0.008, n_estimator = 5000, seed = 42, depth = 10, submit4.csv

'10 fold CV, score: 0.8416127641119365', LR-0.008, n_estimator = 5000, seed = 2020

'10 fold CV, score: 0.8415039416559706', LR-0.008, n_estimator = 5000, seed = 42

### Since we are using cross validation, We'll have `n` different predictions (where n is the number of fold)

In [19]:
oof_prediction = pd.DataFrame(test_oofs).T

In [20]:
oof_prediction.columns = ['fold_'+ str(i) for i in range(1, n_skf + 1)]

In [21]:
oof_prediction.head()

fold_1    fold_2    fold_3  ...   fold_13   fold_14   fold_15
0  0.314114  0.288612  0.342378  ...  0.303312  0.322182  0.304440
1  0.425212  0.332593  0.377943  ...  0.401058  0.397946  0.351174
2  0.354195  0.394853  0.381181  ...  0.356760  0.359079  0.369657
3  0.729698  0.743151  0.763501  ...  0.738476  0.725104  0.738868
4  0.167424  0.158714  0.173079  ...  0.189161  0.174128  0.151464

[5 rows x 15 columns]

In [22]:
sub = pd.read_csv('/content/drive/My Drive/DSN AI Bootcamp Qualification Hackathon/SampleSubmission.csv')

In [ ]:
### We take the mean of all 5 predictions to generate submission

In [23]:
sub['default_status'] = np.mean(test_oofs, axis = 0)

In [24]:
sub.to_csv('submit12.csv', index = False)

In [31]:
len(model.feature_importances_)

50

In [29]:
feature_importance_df

importance
0     7.992341
1     8.205858
2     2.586986
3     1.198259
4     0.702730
5     3.137814
6     1.894769
7     1.890604
8     2.341815
9     1.614640
10    1.856429
11    1.038550
12    1.520874
13    1.498717
14    1.938334
15    1.103915
16    1.460388
17    0.217587
18    1.089553
19    1.035362
20    1.447902
21    1.608628
22    0.996053
23    1.930342
24    2.287043
25    1.433314
26    1.538687
27    1.012848
28    1.662133
29    1.530245
30    0.952576
31    1.594528
32    1.569407
33    1.914958
34    1.342812
35    1.930324
36    2.177405
37    2.112077
38    0.644803
39    1.871662
40    1.159603
41    2.902707
42    2.039873
43    1.542816
44    1.675353
45    0.760150
46   10.690036
47    0.656370
48    1.988022
49    0.701802

In [30]:
len(test.columns)

267

In [28]:
feature_importance_df = pd.DataFrame(model.feature_importances_, columns=['importance'])
feature_importance_df['feature'] = train.columns

plt.figure(figsize=(12, 6));
sns.barplot(x="importance", y="feature", data=feature_importance_df.sort_values(by = ['importance'], ascending = False).head(40))
plt.title('CatBoostClassifier features importance:');

ValueError: ignored

### Suggestions:
- Base your feature selection on the local cv score i.e When you add, remove or engineer feature check `score` (cell 17)
- Increase number of fold to 10, 15. See how score imporve on leaderboard.
- Try differnt value of learning rate. I'll advice anything < 0.01
- Create as many features as possible.
- Find code for Xgboost and lgbm using this method of modelling, or better still tweak this one (mind you, names of parameters may be slightly different.